### PR TITLE
Fix the missing profiling step

### LINKED ISSUE #76: When parsing the trace files the last profiling step is not loaded
### 🐛 Describe the bug

When parsing the trace files, the last profiling step is not loaded. This is related to this line
```
cpu_kernels = cpu_kernels[cpu_kernels["ts"] < last_profiler_start]
```

### Steps to reproduce

Call `analyzer = TraceAnalysis(trace_dir=trace_dir)` to load the trace files and notice that the last profiling step is missing in `analyzer.t.traces[0]`

### Expected behavior

The last profiling step should also be loaded and analyzed.

### Environment

Linux x86_64

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by Juntao-Wang**
## What does this PR do?
Fixes #76 , when filtering the `cpu_kernels` the last profiling step is not included.

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/HolisticTraceAnalysis/blob/main/CHANGELOG.md)?
  - [x] N/A

### CODE REVIEW

File: hta/common/trace.py [hunk 1]
```diff
@@ -706,7 +706,7 @@ def filter_gpu_kernels_for_one_rank(trace_df: pd.DataFrame) -> pd.DataFrame:
             last_profiler_start = cpu_kernels[cpu_kernels["name"].isin(profiler_steps)][
                 "ts"
             ].max()
-            cpu_kernels = cpu_kernels[cpu_kernels["ts"] < last_profiler_start]
+            cpu_kernels = cpu_kernels[cpu_kernels["ts"] <= last_profiler_start]
```
Comment (fengxizhou) [inline hta/common/trace.py:709]: You may change the signature to:
def _filter_irrelevant_gpu_kernels(self, include_last_profiler_step: Optional[bool] = False) -> None:

and then change the code as:
if include_last_profiler_step:
      cpu_kernels = cpu_kernels[cpu_kernels["ts"] <= last_profiler_start]
else:
       cpu_kernels = cpu_kernels[cpu_kernels["ts"] < last_profiler_start]

Comment (fengxizhou) [reply hta/common/trace.py:709]: After the change, you can run 
```
make test
```
to verify that all tests run successfully.

Comment (Juntao-Wang) [reply hta/common/trace.py]: Thanks for the suggestions! I have added the propagated flag to TraceAnalysis class and also added the corresponding test cases. Related test cases are all passed, but I noticed that one other test case failed, which I believe is unrelated to the changes made here. I have pasted the fail case below:
```
======================================================================
FAIL: test_create_rank_to_trace_dict_with_mixed_dir (tests.test_trace_file.TestTraceFile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/juntao/github/HolisticTraceAnalysis/tests/test_trace_file.py", line 78, in test_create_rank_to_trace_dict_with_mixed_dir
    self.assertEqual(
AssertionError: Tuples differ: (True, {0: 'tests/data/mixed_files/rank-0.json.gz'}) != (True, {0: 'tests/data/mixed_files/rank_non_gpu.json.gz'})

First differing element 1:
{0: 'tests/data/mixed_files/rank-0.json.gz'}
{0: 'tests/data/mixed_files/rank_non_gpu.json.gz'}

- (True, {0: 'tests/data/mixed_files/rank-0.json.gz'})
?                                        ^^

+ (True, {0: 'tests/data/mixed_files/rank_non_gpu.json.gz'})
?                                        ^^^^^^^^
```

**Comment by anupambhatnagar** [comment]
@Juntao-Wang please run pre-commit locally and then make another commit so that the pre-commit checks can pass. Here are the instructions to run pre-commit locally. https://github.com/facebookresearch/HolisticTraceAnalysis/blob/main/CONTRIBUTING.md#pre-commit-recommended

**Comment by anupambhatnagar** [comment]
Instead of modifying the existing test, we need to add another test where the boolean flag that you are creating is set to True. Please add that test and revert the changes to the existing test.

**Comment by anupambhatnagar** [comment]
@Juntao-Wang one more thing for this PR to be approved we need to run the tests in our CI. The tests will be automatically triggered if you create a branch from facebookresearch:main and not from your fork. If any tests fail, I can help fix them.

**Comment by Juntao-Wang** [comment]
@anupambhatnagar Thanks for the instructions! I have fixed the pre-commit and the test cases, but I am not sure how to create a branch from facebookresearch:main instead of my own fork and would like to kindly ask for further help on this step. Many thanks!

**Comment by anupambhatnagar** [comment]
> @anupambhatnagar Thanks for the instructions! I have fixed the pre-commit and the test cases, but I am not sure how to create a branch from facebookresearch:main instead of my own fork and would like to kindly ask for further help on this step. Many thanks!

Clone the repo locally on your machine (using `git clone https://github.com/facebookresearch/HolisticTraceAnalysis.git`) and create a new branch (e.g., `add_last_profiler_step`) there. Once you have transferred the changes from your fork to this branch then commit and push.

File: tests/test_trace_analysis.py [hunk 2]
```diff
@@ -34,6 +35,9 @@ def setUpClass(cls):
         rank_non_gpu_trace_dir: str = "tests/data/rank_non_gpu/"
         h100_trace_dir: str = "tests/data/h100"
         cls.vision_transformer_t = TraceAnalysis(trace_dir=vision_transformer_trace_dir)
+        cls.vision_transformer_include_last_profiler_step_t = TraceAnalysis(
+            trace_dir=vision_transformer_trace_dir, include_last_profiler_step=True
```
[the remark this answers is not in the document]
Comment (Juntao-Wang) [reply tests/test_trace_analysis.py]: Changes have been incorporated, they are now moved to the new test.

**Comment by Juntao-Wang** [comment]
> > @anupambhatnagar Thanks for the instructions! I have fixed the pre-commit and the test cases, but I am not sure how to create a branch from facebookresearch:main instead of my own fork and would like to kindly ask for further help on this step. Many thanks!
> 
> Clone the repo locally on your machine (using `git clone https://github.com/facebookresearch/HolisticTraceAnalysis.git`) and create a new branch (e.g., `add_last_profiler_step`) there. Once you have transferred the changes from your fork to this branch then commit and push.

Many thanks! I was trying to push the new branch to `facebookresearch/HolisticTraceAnalysis.git`, however I got `Permission to facebookresearch/HolisticTraceAnalysis.git denied`

**Comment by anupambhatnagar** [comment]
@Juntao-Wang I have given you write permissions on the repo so you should be able to push now. I would recommend closing this PR and creating a new one which does the push from the branch on the cloned repo not the forked repo (this PR  is from the forked repo). 

You should have received an email which will mention that you have write permissions to this repo. Please accept the terms of that email before you try to push the changes. In case there are still some issues, i'm happy to help.
Thanks for your contribution!

**Comment by Juntao-Wang** [comment]
Great! Many thanks! I will close this PR and open a new one.

**Pull request closed without merging**